android/src/main/java/com/polidea/blemulator/SimulatedAdapter.java

```diff
@@ -302,30 +312,49 @@ public void onSuccess(DeviceContainer deviceContainer) {
             throw new BleError(BleErrorCode.DeviceNotConnected, "Device not connected", 0);
         }
 
-        if (knownPeripherals.get(deviceIdentifier).getCharacteristics() == null) {
+        if (knownPeripherals.get(deviceIdentifier).getCharacteristicContainersIndexedByServiceUuids() == null) {
             throw new BleError(BleErrorCode.CharacteristicsNotDiscovered, "Discovery not done on this device", 0);
         }
 
-        return knownPeripherals
+        List<CharacteristicContainer> characteristicContainers = knownPeripherals
                 .get(deviceIdentifier)
-                .getCharacteristics()
-                .get(serviceUUID)
-                .toArray(new Characteristic[0]);
+                .getCharacteristicContainersIndexedByServiceUuids()
+                .get(serviceUUID);
+
+        List<Characteristic> characteristics = new LinkedList<>();
+        if (characteristicContainers != null) {
+            for (CharacteristicContainer characteristicContainer : characteristicContainers) {
+                characteristics.add(characteristicContainer.getCharacteristic());
+            }
+        }
+
+        return characteristics;
     }
 
     @Override
-    public Characteristic[] getCharacteristicsForService(int serviceIdentifier) throws BleError {
+    public List<Characteristic> getCharacteristicsForService(int serviceIdentifier) throws BleError {
         Log.i(TAG, "getCharacteristicForService");
 
 
         for (DeviceContainer deviceContainer : knownPeripherals.values()) {
             if (deviceContainer.getServices() != null) {
                 for (Service service : deviceContainer.getServices()) {
                     if (service.getId() == serviceIdentifier) {
-                        return deviceContainer
-                                .getCharacteristics()
-                                .get(service.getUuid().toString().toUpperCase())
-                                .toArray(new Characteristic[0]);
+                        if (!deviceContainer.isConnected()) {
+                            throw new BleError(BleErrorCode.DeviceNotConnected, "Device not connected", 0);
+                        }
+                        List<CharacteristicContainer> characteristicContainers = deviceContainer
+                                .getCharacteristicContainersIndexedByServiceUuids()
+                                .get(service.getUuid().toString().toUpperCase());
```

Comment:
I wonder if `.toUpperCase()` sanitisation should be made at the very first moment the library takes over input values?